# ETL Pipeline Preparation
Follow the instructions below to help you create your ETL pipeline.
### 1. Import libraries and load datasets.
- Import Python libraries
- Load `messages.csv` into a dataframe and inspect the first few lines.
- Load `categories.csv` into a dataframe and inspect the first few lines.

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [38]:
# load messages dataset
messages = pd.read_csv('../data/disaster_messages.csv')
messages.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  
0  Un front froid se retrouve sur Cuba ce matin. ...  direct  
1                 Cyclone nan fini osinon li pa fini  direct  
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct  
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct  
4  facade ouest d Haiti et le reste du pays aujou...  direct

In [39]:
messages.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26248 entries, 0 to 26247
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        26248 non-null  int64 
 1   message   26248 non-null  object
 2   original  10184 non-null  object
 3   genre     26248 non-null  object
dtypes: int64(1), object(3)
memory usage: 820.4+ KB


In [40]:
# load categories dataset
categories = pd.read_csv('../data/disaster_categories.csv')
categories.head()

id                                         categories
0   2  related-1;request-0;offer-0;aid_related-0;medi...
1   7  related-1;request-0;offer-0;aid_related-1;medi...
2   8  related-1;request-0;offer-0;aid_related-0;medi...
3   9  related-1;request-1;offer-0;aid_related-1;medi...
4  12  related-1;request-0;offer-0;aid_related-0;medi...

In [41]:
categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26248 entries, 0 to 26247
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          26248 non-null  int64 
 1   categories  26248 non-null  object
dtypes: int64(1), object(1)
memory usage: 410.2+ KB


In [42]:
# Checking if 'id' columns match
(categories['id'] != messages['id']).sum()

0

### 2. Merge datasets.
- Merge the messages and categories datasets using the common id
- Assign this combined dataset to `df`, which will be cleaned in the following steps

In [43]:
# merge datasets
df = messages.merge(categories, how='inner', on='id')
df.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  \
0  Un front froid se retrouve sur Cuba ce matin. ...  direct   
1                 Cyclone nan fini osinon li pa fini  direct   
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct   
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct   
4  facade ouest d Haiti et le reste du pays aujou...  direct   

                                          categories  
0  related-1;request-0;offer-0;aid_related-0;medi...  
1  related-1;request-0;offer-0;aid_related-1;medi...  
2  related-1;request-0;offer-0;aid_related-0;medi...  
3  related-1;request-1;offer-0;aid_related-1;medi...  
4  related-1;request-0;offer-0;aid_related-0;medi...

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26386 entries, 0 to 26385
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          26386 non-null  int64 
 1   message     26386 non-null  object
 2   original    10246 non-null  object
 3   genre       26386 non-null  object
 4   categories  26386 non-null  object
dtypes: int64(1), object(4)
memory usage: 1.2+ MB


#### Why are there more entries after merging? Are there duplicates?

In [45]:
df.duplicated().sum()

170

170 Entries are double. I need to take care of this downstream.

### 3. Split `categories` into separate category columns.
- Split the values in the `categories` column on the `;` character so that each value becomes a separate column. You'll find [this method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.Series.str.split.html) very helpful! Make sure to set `expand=True`.
- Use the first row of categories dataframe to create column names for the categories data.
- Rename columns of `categories` with new column names.

In [46]:
# create a dataframe of the 36 individual category columns
categories = categories['categories'].str.split(';', expand = True)
cols = categories.iloc[0].str.split('-').str[0]
categories.columns = cols
categories.head()

0    related    request    offer    aid_related    medical_help  \
0  related-1  request-0  offer-0  aid_related-0  medical_help-0   
1  related-1  request-0  offer-0  aid_related-1  medical_help-0   
2  related-1  request-0  offer-0  aid_related-0  medical_help-0   
3  related-1  request-1  offer-0  aid_related-1  medical_help-0   
4  related-1  request-0  offer-0  aid_related-0  medical_help-0   

0    medical_products    search_and_rescue    security    military  \
0  medical_products-0  search_and_rescue-0  security-0  military-0   
1  medical_products-0  search_and_rescue-0  security-0  military-0   
2  medical_products-0  search_and_rescue-0  security-0  military-0   
3  medical_products-1  search_and_rescue-0  security-0  military-0   
4  medical_products-0  search_and_rescue-0  security-0  military-0   

0    child_alone  ...    aid_centers    other_infrastructure  \
0  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
1  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
2  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
3  child_alone-0  ...  aid_centers-0  other_infrastructure-0   
4  child_alone-0  ...  aid_centers-0  other_infrastructure-0   

0    weather_related    floods    storm    fire    earthquake    cold  \
0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
1  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0   
2  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
3  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
4  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   

0    other_weather    direct_report  
0  other_weather-0  direct_report-0  
1  other_weather-0  direct_report-0  
2  other_weather-0  direct_report-0  
3  other_weather-0  direct_report-0  
4  other_weather-0  direct_report-0  

[5 rows x 36 columns]

### 4. Convert category values to just numbers 0 or 1.
- Iterate through the category columns in df to keep only the last character of each string (the 1 or 0). For example, `related-0` becomes `0`, `related-1` becomes `1`. Convert the string to a numeric value.
- You can perform [normal string actions on Pandas Series](https://pandas.pydata.org/pandas-docs/stable/text.html#indexing-with-str), like indexing, by including `.str` after the Series. You may need to first convert the Series to be of type string, which you can do with `astype(str)`.

In [47]:
for column in categories:
    # set each value to be the last character of the string
    categories[column] = categories[column].str[-1].astype(int)
    
    # convert column from string to numeric
    #categories[column] = categories[column].astype(int)
categories.head()

0  related  request  offer  aid_related  medical_help  medical_products  \
0        1        0      0            0             0                 0   
1        1        0      0            1             0                 0   
2        1        0      0            0             0                 0   
3        1        1      0            1             0                 1   
4        1        0      0            0             0                 0   

0  search_and_rescue  security  military  child_alone  ...  aid_centers  \
0                  0         0         0            0  ...            0   
1                  0         0         0            0  ...            0   
2                  0         0         0            0  ...            0   
3                  0         0         0            0  ...            0   
4                  0         0         0            0  ...            0   

0  other_infrastructure  weather_related  floods  storm  fire  earthquake  \
0                     0                0       0      0     0           0   
1                     0                1       0      1     0           0   
2                     0                0       0      0     0           0   
3                     0                0       0      0     0           0   
4                     0                0       0      0     0           0   

0  cold  other_weather  direct_report  
0     0              0              0  
1     0              0              0  
2     0              0              0  
3     0              0              0  
4     0              0              0  

[5 rows x 36 columns]

### 5. Replace `categories` column in `df` with new category columns.
- Drop the categories column from the df dataframe since it is no longer needed.
- Concatenate df and categories data frames.

In [48]:
categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26248 entries, 0 to 26247
Data columns (total 36 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   related                 26248 non-null  int32
 1   request                 26248 non-null  int32
 2   offer                   26248 non-null  int32
 3   aid_related             26248 non-null  int32
 4   medical_help            26248 non-null  int32
 5   medical_products        26248 non-null  int32
 6   search_and_rescue       26248 non-null  int32
 7   security                26248 non-null  int32
 8   military                26248 non-null  int32
 9   child_alone             26248 non-null  int32
 10  water                   26248 non-null  int32
 11  food                    26248 non-null  int32
 12  shelter                 26248 non-null  int32
 13  clothing                26248 non-null  int32
 14  money                   26248 non-null  int32
 15  missing_people     

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26386 entries, 0 to 26385
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          26386 non-null  int64 
 1   message     26386 non-null  object
 2   original    10246 non-null  object
 3   genre       26386 non-null  object
 4   categories  26386 non-null  object
dtypes: int64(1), object(4)
memory usage: 1.2+ MB


In [50]:
# concatenate the original dataframe with the new `categories` dataframe
df = pd.concat([df, categories], axis = 1)
df.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  \
0  Un front froid se retrouve sur Cuba ce matin. ...  direct   
1                 Cyclone nan fini osinon li pa fini  direct   
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct   
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct   
4  facade ouest d Haiti et le reste du pays aujou...  direct   

                                          categories  related  request  offer  \
0  related-1;request-0;offer-0;aid_related-0;medi...      1.0      0.0    0.0   
1  related-1;request-0;offer-0;aid_related-1;medi...      1.0      0.0    0.0   
2  related-1;request-0;offer-0;aid_related-0;medi...      1.0      0.0    0.0   
3  related-1;request-1;offer-0;aid_related-1;medi...      1.0      1.0    0.0   
4  related-1;request-0;offer-0;aid_related-0;medi...      1.0      0.0    0.0   

   aid_related  medical_help  ...  aid_centers  other_infrastructure  \
0          0.0           0.0  ...          0.0                   0.0   
1          1.0           0.0  ...          0.0                   0.0   
2          0.0           0.0  ...          0.0                   0.0   
3          1.0           0.0  ...          0.0                   0.0   
4          0.0           0.0  ...          0.0                   0.0   

   weather_related  floods  storm  fire  earthquake  cold  other_weather  \
0              0.0     0.0    0.0   0.0         0.0   0.0            0.0   
1              1.0     0.0    1.0   0.0         0.0   0.0            0.0   
2              0.0     0.0    0.0   0.0         0.0   0.0            0.0   
3              0.0     0.0    0.0   0.0         0.0   0.0            0.0   
4              0.0     0.0    0.0   0.0         0.0   0.0            0.0   

   direct_report  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  

[5 rows x 41 columns]

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26386 entries, 0 to 26385
Data columns (total 41 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      26386 non-null  int64  
 1   message                 26386 non-null  object 
 2   original                10246 non-null  object 
 3   genre                   26386 non-null  object 
 4   categories              26386 non-null  object 
 5   related                 26248 non-null  float64
 6   request                 26248 non-null  float64
 7   offer                   26248 non-null  float64
 8   aid_related             26248 non-null  float64
 9   medical_help            26248 non-null  float64
 10  medical_products        26248 non-null  float64
 11  search_and_rescue       26248 non-null  float64
 12  security                26248 non-null  float64
 13  military                26248 non-null  float64
 14  child_alone             26248 non-null

Two issues:

1) 138 messages don't have flagged categories => I will remove these

2) The category flag format switched from int to float for some reason. => I will re-format

In [52]:
# Checking examples where information is missing in the categories
df[df['related'].isna()]

id                                            message original  \
26248  30112  The 1 July meeting of the Support and Follow-u...      NaN   
26249  30113  Japan's overseas missions are accepting relief...      NaN   
26250  30114  According to officials, Kabul River and Swat R...      NaN   
26251  30115  The gross relief food requirements for June-De...      NaN   
26252  30116  Authorities have built tent compounds in flatt...      NaN   
...      ...                                                ...      ...   
26381  30261  The training demonstrated how to enhance micro...      NaN   
26382  30262  A suitable candidate has been selected and OCH...      NaN   
26383  30263  Proshika, operating in Cox's Bazar municipalit...      NaN   
26384  30264  Some 2,000 women protesting against the conduc...      NaN   
26385  30265  A radical shift in thinking came about as a re...      NaN   

      genre                                         categories  related  \
26248  news  related-0;request-0;offer-0;aid_related-0;medi...      NaN   
26249  news  related-0;request-0;offer-0;aid_related-0;medi...      NaN   
26250  news  related-1;request-0;offer-0;aid_related-0;medi...      NaN   
26251  news  related-1;request-0;offer-0;aid_related-1;medi...      NaN   
26252  news  related-1;request-0;offer-0;aid_related-1;medi...      NaN   
...     ...                                                ...      ...   
26381  news  related-0;request-0;offer-0;aid_related-0;medi...      NaN   
26382  news  related-0;request-0;offer-0;aid_related-0;medi...      NaN   
26383  news  related-1;request-0;offer-0;aid_related-0;medi...      NaN   
26384  news  related-1;request-0;offer-0;aid_related-1;medi...      NaN   
26385  news  related-1;request-0;offer-0;aid_related-0;medi...      NaN   

       request  offer  aid_related  medical_help  ...  aid_centers  \
26248      NaN    NaN          NaN           NaN  ...          NaN   
26249      NaN    NaN          NaN           NaN  ...          NaN   
26250      NaN    NaN          NaN           NaN  ...          NaN   
26251      NaN    NaN          NaN           NaN  ...          NaN   
26252      NaN    NaN          NaN           NaN  ...          NaN   
...        ...    ...          ...           ...  ...          ...   
26381      NaN    NaN          NaN           NaN  ...          NaN   
26382      NaN    NaN          NaN           NaN  ...          NaN   
26383      NaN    NaN          NaN           NaN  ...          NaN   
26384      NaN    NaN          NaN           NaN  ...          NaN   
26385      NaN    NaN          NaN           NaN  ...          NaN   

       other_infrastructure  weather_related  floods  storm  fire  earthquake  \
26248                   NaN              NaN     NaN    NaN   NaN         NaN   
26249                   NaN              NaN     NaN    NaN   NaN         NaN   
26250                   NaN              NaN     NaN    NaN   NaN         NaN   
26251                   NaN              NaN     NaN    NaN   NaN         NaN   
26252                   NaN              NaN     NaN    NaN   NaN         NaN   
...                     ...              ...     ...    ...   ...         ...   
26381                   NaN              NaN     NaN    NaN   NaN         NaN   
26382                   NaN              NaN     NaN    NaN   NaN         NaN   
26383                   NaN              NaN     NaN    NaN   NaN         NaN   
26384                   NaN              NaN     NaN    NaN   NaN         NaN   
26385                   NaN              NaN     NaN    NaN   NaN         NaN   

       cold  other_weather  direct_report  
26248   NaN            NaN            NaN  
26249   NaN            NaN            NaN  
26250   NaN            NaN            NaN  
26251   NaN            NaN            NaN  
26252   NaN            NaN            NaN  
...     ...            ...            ...  
26381   NaN            NaN            NaN  
26382   NaN            NaN    

In [53]:
# Dropping all rows where categories have no entry
df = df[df['related'].notna()]

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26248 entries, 0 to 26247
Data columns (total 41 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      26248 non-null  int64  
 1   message                 26248 non-null  object 
 2   original                10246 non-null  object 
 3   genre                   26248 non-null  object 
 4   categories              26248 non-null  object 
 5   related                 26248 non-null  float64
 6   request                 26248 non-null  float64
 7   offer                   26248 non-null  float64
 8   aid_related             26248 non-null  float64
 9   medical_help            26248 non-null  float64
 10  medical_products        26248 non-null  float64
 11  search_and_rescue       26248 non-null  float64
 12  security                26248 non-null  float64
 13  military                26248 non-null  float64
 14  child_alone             26248 non-null

Looks good, now formatting

In [55]:
df[categories.columns] = df[categories.columns].astype(int)

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26248 entries, 0 to 26247
Data columns (total 41 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   id                      26248 non-null  int64 
 1   message                 26248 non-null  object
 2   original                10246 non-null  object
 3   genre                   26248 non-null  object
 4   categories              26248 non-null  object
 5   related                 26248 non-null  int32 
 6   request                 26248 non-null  int32 
 7   offer                   26248 non-null  int32 
 8   aid_related             26248 non-null  int32 
 9   medical_help            26248 non-null  int32 
 10  medical_products        26248 non-null  int32 
 11  search_and_rescue       26248 non-null  int32 
 12  security                26248 non-null  int32 
 13  military                26248 non-null  int32 
 14  child_alone             26248 non-null  int32 
 15  wa

Looks good

### 6. Remove duplicates.
- Check how many duplicates are in this dataset.
- Drop the duplicates.
- Confirm duplicates were removed.

In [57]:
# check number of duplicates
df.duplicated().sum()

26

We knwe about duplicates, still some left

In [58]:
# drop duplicates
df.drop_duplicates(inplace = True)

In [59]:
# check number of duplicates
df.duplicated().sum()

0

### 7. Save the clean dataset into an sqlite database.
You can do this with pandas [`to_sql` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_sql.html) combined with the SQLAlchemy library. Remember to import SQLAlchemy's `create_engine` in the first cell of this notebook to use it below.

In [60]:
engine = create_engine('sqlite:///../data/disaster_responses.db', encoding="UTF-8")
df.to_sql('disaster_messages', engine, index=False)

### 8. Use this notebook to complete `etl_pipeline.py`
Use the template file attached in the Resources folder to write a script that runs the steps above to create a database based on new datasets specified by the user. Alternatively, you can complete `etl_pipeline.py` in the classroom on the `Project Workspace IDE` coming later.